### PR TITLE
[Examples] Add Sampling Examples

### PR DESCRIPTION
**Pull request by keli-wen**
## Description

I added a folder named `examples/sampling`, and within it, I included a file called `node_classification.py`. I also provided highlighted comments for APIs in the file that are associated with the sampling operation in the DGL library.

The execution log:
```
$ python node_classification.py
Training in mixed mode.
Loading data
Training...
Epoch 00000 | Loss 1.3346 | Accuracy 0.8561 
Epoch 00001 | Loss 0.5894 | Accuracy 0.8772 
Epoch 00002 | Loss 0.4901 | Accuracy 0.8855 
Epoch 00003 | Loss 0.4463 | Accuracy 0.8893 
Epoch 00004 | Loss 0.4223 | Accuracy 0.8931 
Epoch 00005 | Loss 0.3992 | Accuracy 0.8964 
Epoch 00006 | Loss 0.3843 | Accuracy 0.8968 
Epoch 00007 | Loss 0.3794 | Accuracy 0.8992 
Epoch 00008 | Loss 0.3699 | Accuracy 0.8987 
Epoch 00009 | Loss 0.3614 | Accuracy 0.9000 
Testing...
100%|██████████████████████████████████| 598/598 [00:33<00:00, 17.86it/s]
100%|██████████████████████████████████| 598/598 [00:33<00:00, 17.67it/s]
100%|██████████████████████████████████| 598/598 [00:33<00:00, 17.74it/s]
Test Accuracy 0.7702
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sampling/node_classification.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+from dgl.dataloading import (
+    DataLoader,
+    MultiLayerFullNeighborSampler,
+    NeighborSampler,
+)
+from ogb.nodeproppred import DglNodePropPredDataset
+
+
+class SAGE(nn.Module):
+    def __init__(self, in_size, hid_size, out_size):
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:25]: hid_size -> hidden_size

There is not benefit to shorten the word, but creates 2 seconds of confusion.

File: examples/sampling/node_classification.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
+"""
+If you want a deeper understanding of node classification. You can
+read the example in the  ``examples/pytorch/graphsage/node_classification.py``
```
Comment (mufeili) [inline examples/sampling/node_classification.py:3]:
```suggestion
read the example in the ``examples/pytorch/graphsage/node_classification.py``
```

File: examples/sampling/node_classification.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
+"""
```
Comment (mufeili) [inline examples/sampling/node_classification.py:1]: Add a sentence describing what this script does. For example, node classification with neighbor sampling.

Comment (keli-wen) [reply examples/sampling/node_classification.py:1]: Get~

File: examples/sampling/node_classification.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            f"Accuracy {acc.item():.4f} "
+        )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--mode",
+        default="mixed",
+        choices=["cpu", "mixed", "puregpu"],
```
Comment (mufeili) [inline examples/sampling/node_classification.py:209]: Consider renaming `puregpu` to `gpu` just like you are not using `purecpu`.

File: examples/sampling/node_classification.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+"""
+
+import argparse
+
+import dgl
+import dgl.nn as dglnn
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torchmetrics.functional as MF
```
Comment (mufeili) [inline examples/sampling/node_classification.py:13]: Consider including a README file to specify the version of the dependencies like `torchmetrics`.

File: examples/sampling/node_classification.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    use_uva = args.mode == "mixed"
+    train_dataloader = DataLoader(
+        g,
+        train_idx,
+        sampler,
+        device=device,
+        batch_size=1024,
+        shuffle=True,
+        drop_last=False,
+        num_workers=0,
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:165]: worth to callout why `num_workers=0`

File: examples/sampling/node_classification.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
+"""
+If you want a deeper understanding of node classification. You can
```
Comment (mufeili) [inline examples/sampling/node_classification.py:2]: Consider referencing the GraphSAGE paper for neighbor sampling.

File: examples/sampling/node_classification.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            num_workers=0,
+        )
+        buffer_device = torch.device("cpu")
+        pin_memory = buffer_device != device
+
+        for layer_idx, layer in enumerate(self.layers):
+            is_last_layer = layer_idx == len(self.layers) - 1
+            y = torch.empty(
+                g.num_nodes(),
+                self.hid_size if (not is_last_layer) else self.out_size,
```
Comment (mufeili) [inline examples/sampling/node_classification.py:84]:
```suggestion
                self.out_size if is_last_layer else self.hid_size,
```
I tend to find this more readable.

File: examples/sampling/node_classification.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+            torch.arange(g.num_nodes()).to(g.device),
+            sampler,
+            device=device,
+            batch_size=batch_size,
+            shuffle=False,
+            drop_last=False,
+            num_workers=0,
+        )
+        buffer_device = torch.device("cpu")
+        pin_memory = buffer_device != device
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:78]: why we use pin_memory in inference but not training?

File: examples/sampling/node_classification.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        use_uva=use_uva,
+    )
+
+    val_dataloader = DataLoader(
+        g,
+        val_idx,
+        sampler,
+        device=device,
+        batch_size=1024,
+        shuffle=True,
```
Comment (mufeili) [inline examples/sampling/node_classification.py:175]: No need to shuffle for validation

File: examples/sampling/node_classification.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+                if layer_idx != len(self.layers) - 1:
+                    h = F.relu(h)
+                    h = self.dropout(h)
+                # By design, our output nodes are contiguous.
+                y[output_nodes[0] : output_nodes[-1] + 1] = h.to(buffer_device)
+            feat = y
+        return y
+
+
+def evaluate(model, graph, dataloader, num_classes):
```
Comment (mufeili) [inline examples/sampling/node_classification.py:101]: Maybe easier to decorate this function with `@torch.no_grad()`

**Review by mufeili — COMMENTED** [review]
done a pass

File: examples/sampling/node_classification.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+        # Three-layer GraphSAGE-mean.
+        self.layers.append(dglnn.SAGEConv(in_size, hid_size, "mean"))
+        self.layers.append(dglnn.SAGEConv(hid_size, hid_size, "mean"))
+        self.layers.append(dglnn.SAGEConv(hid_size, out_size, "mean"))
+        self.dropout = nn.Dropout(0.5)
+        self.hid_size = hid_size
+        self.out_size = out_size
+
+    def forward(self, blocks, x):
+        h = x
```
[the remark this answers is not in the document]
Comment (keli-wen) [reply examples/sampling/node_classification.py:37]: In the code, `h = x` is for clarity and safety. `h` signifies the evolving **`state`** through the layers. Modifying `h` avoids changing `x` directly, preventing unexpected side effects. Perhaps naming it `hidden_x` would make it more readable?

File: examples/sampling/node_classification.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+        # This sampler is used in the Graph Neural Networks (GNN) training
+        # process to provide neighbor sampling, which is crucial for
+        # efficient training of GNN on large graphs.
+        #
+        # The first argument '1' indicates the number of layers for
+        # the neighbor sampling. In this case, it's set to 1, meaning
+        # only the direct neighbors of each node will be included in the
+        # sampling.
+        #
+        # The 'prefetch_node_feats' parameter specifies the node features
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:88]: With these comment, it is still not very clear how prefetch_node_feats happens, even with this doc (https://docs.dgl.ai/en/latest/guide/minibatch-prefetching.html).

@BarclayII can you chime in here?

File: examples/sampling/node_classification.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+    # to be sampled at each layer. Here, we have three layers, and
+    # 10 neighbors will be randomly selected for each node at each
+    # layer.
+    #
+    # The 'prefetch_node_feats' and 'prefetch_labels' parameters
+    # specify the node features and labels that need to be pre-fetched
+    # during sampling. Pre-fetching data in this way is advantageous
+    # as it can reduce the I/O overhead during the training process,
+    # making the whole computation more efficient. In this case,
+    # the feature 'feat' and the label 'label' will be pre-fetched.
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:177]:
```
Pre-fetching data in this way is advantageous
    # as it can reduce the I/O overhead during the training process,
    # making the whole computation more efficient. In this case,
    # the feature 'feat' and the label 'label' will be pre-fetched.
```

This part does not add enough information as I can easily get the point that the prefetch will reduce I/O overhead in someway, but it is more helpful to elaborate how prefetching works.

File: examples/sampling/node_classification.py (hunk 13 above)
Comment (keli-wen) [reply examples/sampling/node_classification.py:88]: In the following text, the `prefetch` parameter is also used, and I've added some advantages of using `prefetch` there. I've done some research and found this blog post to be quite detailed: https://discuss.dgl.ai/t/confusion-about-feature-prefetch/3573/4. In summary, it appears to be a technique used to speed up I/O operations.

Comment (frozenbugs) [reply examples/sampling/node_classification.py:88]: https://docs.google.com/document/d/1uyUfRc9SuavR-_Nr-Rn4acEehD6cUE5oTjXHPmUkfWI/edit#bookmark=id.yp7uomdejg6w

Here is the explanation of prefetch, can you update the documentation accordingly?

File: examples/sampling/node_classification.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+    train_dataloader = DataLoader(
+        g,
+        train_idx,
+        sampler,
+        device=device,
+        batch_size=1024,
+        shuffle=True,
+        drop_last=False,
+        # If `g` is on gpu or `use_uva` is True,
+        # `num_workers` must be zero else it will cause error.
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:194]: nit: Always make full use of 1 line (max 80 chars) before change to new line.
```suggestion
        # If `g` is on gpu or `use_uva` is True, `num_workers` must be zero,
        # otherwise it will cause error.
```

File: examples/sampling/node_classification.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+large graphs using efficient neighbor sampling.
+
+Paper: [Inductive Representation Learning on Large Graphs]
+(https://arxiv.org/abs/1706.02216)
+
+If you want a deeper understanding of node classification. You can
+read the example in the `examples/pytorch/graphsage/node_classification.py`
+TODO(#5797): Move `graphsage/node_classification.py` to the `examples/core/`.
+
+This flowchart describes the main functional sequence of the provided example.
```
Comment (mufeili) [inline examples/sampling/node_classification.py:12]: Do you still need this flowchart with the abstraction in `if __name__ == "__main__"`?

**Review by mufeili — APPROVED** [review]
I'm mostly good. Thank you for the great efforts.

File: examples/sampling/node_classification.py [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+        total_loss = 0
+        # A block is a graph consisting of two sets of nodes: the
+        # source nodes and destination nodes. The source and destination
+        # nodes can have multiple node types. All the edges connect from
+        # source nodes to destination nodes.
+        # For more details: https://discuss.dgl.ai/t/what-is-the-block/2932.
+        for it, (input_nodes, output_nodes, blocks) in enumerate(
+            train_dataloader
+        ):
+            # The input features from the source nodes in the
```
Comment (frozenbugs) [inline examples/sampling/node_classification.py:225]: ditto

File: examples/sampling/node_classification.py (hunk 16 above)
Comment (frozenbugs) [reply examples/sampling/node_classification.py:12]: I find it helpful since it has a  deeper than the main function.

**Comment by frozenbugs** [comment]
@caojy1998 please also take a look with your fresh eyes.

**Comment by mufeili** [comment]
@dgl-bot

**Review by frozenbugs — APPROVED** [review]
